src/human_encoding/README.md

```diff
@@ -0,0 +1,189 @@
+yet to be defined. Holes have a different namespace than other names.
+
+A VALUE is one of
+* the literal `_`, which is interpreted as the empty value;
+* a binary literal `0b[01]+` which is interpreted as a sequence of bits; or
+* a hex literal `0x[0-9a-f]+` which is interpreted as a sequence of 4-bit nybbles
+
+An ENTROPY is a VALUE whose size is between 128 and 512 bits inclusive. Internally
+it is 0-padded out to 512 bits.
+
+Finally, TYPE is
+
+* a literal `_`, indicating no type bound;
+* a NAME;
+* a literal `1`, indicating the unit type;
+* a literal `2`, indicating the bit type;
+* `2^n`, where `n` is any power of two, in decimal with no spaces or punctuation;
+* `(` followed by another TYPE followed by `)`;
+* another TYPE, followed by `+`, followed by another TYPE; or
+* another TYPE, followed by `*`, followed by another TYPE.
+
+Here `*` has higher precedence than `+`, and both `+` and `*` are left-associative.
+
+## Semantics: Definitions
+
+As above, Simplicity expressions are a series of **definitions** of the form
+
+    NAME := EXPRESSION :: TYPE -> TYPE
+
+We refer to the `NAME` part as the **name**, `EXPRESSION` as the **expression**,
+and `TYPE -> TYPE` as the **type ascription**. If such a definition appears anywhere
+in a Simplicity encoding, then whenever that name appears in the expression of
+any other definition, its expression is substituted in place of it.
+
+Either the expression or type ascription may be omitted, but not both. Omitting
+the type ascription is equivalent to providing the ascription `_ -> _`; in both
+cases this indicates that no type hint should be given to the assembler.
+
+If the expression is omitted, this definition only provides a type hint. For a
+given name, it is permissible to have multiple definitions, but exactly one must
+provide an expression. This allows the user to provide any number of type-hint-only
+definitions which may be helpful in clarifying a program.
+
+For example, in the encoding
+
+    node := unit :: _ -> 1
+    main := comp node node
+
+the name `node` is substituted by `unit` both places that it appears. We can see
+that starting from the name `main`, by recursive substitution, we obtain a single
+Simplicity expression. The type checker will ensure that the target type of `node`
+is 1.
+
+In general, we do not need to obtain a single expression. It is permissible to
+encode a "DAG forest".
+
+The name `main` is special for several reasons:
+* An expression named `name` implicitly has the type ascription `1 -> 1`. That is, it must always be a program.
+* To generate a commitment-time program from an expression, it must be called `main`.
+* Type ascriptions for `main` and its children are enforced **after** type inference has completed, so they act as sanity checks but cannot change the output of type inference. For other expressions, type ascriptions are enforced **before** and may guide inference.
```

Comment:
aed42d0045f0403f05e263c737bf6dc3c4153d9a: We should define what a Simplicity program is, and also what a (semantic) Simplicity expression is.

Reply:
Yes, but probably not in the README for the human-readable encoding.